### PR TITLE
Combine solution injection systems; Fix embeddable injectors

### LINKED ISSUE #20708: Arrows inject chems when they shouldn't
## Description
If you throw an arrow at a person or whoever, it will inject it's chems into them, while not hitting them and flying past.
Probably shouldn't be like that since it doesn't actually hit?

**Reproduction**
1. Put poison on arrow
2. Throw at a person
3. Arrow flies past
4. Person still got injected

### LINKED ISSUE #26233: Embeddable injectors aren't injecting
## Description

 When items like the hypodart, normal dart, and spear are thrown into people, they don't inject their contents.
The hypodart and spear still inject on melee swing.
**Reproduction**

Get hypodart
Fill with amatoxin
Throw at Urist
No poison damage, amatoxin is still in dart
**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by Tayrtahn**
## About the PR

Reduced code duplication between injection systems and fixed some bugs.

## Why / Balance / Technical details

After fixing injections from reflected projectiles in #26141, it was reported that embeddable injectors weren't working (#26233).

This gave me the idea to split SolutionInjectOnCollide into two components/systems to handle projectiles and embeds separately. I realized that there would be some gross code reuse, so I switched to making a single SolutionInjectOnEventSystem that handles behavior for SolutionInjectOnProjectileHitComponent and SolutionInjectOnEmbedComponent, using BaseSolutionInjectOnEventComponent as a common base. This also removes the need to specify FixtureId, since the collisions are processed by the events.

While implementing that, I noticed that MeleeChemicalInjector was nearly identical to SolutionInjectOnCollide, for both the component and the system. So I refactored that to also inherit from BaseSolutionInjectOnEventComponent and go through SolutionInjectOnEventSystem. Code duplication begone!

This also gives both systems features that were previously added to one but not the other. Hardsuits can now block injections from projectiles without the PierceArmor field set to true (though the implementation is still ugly and should be reworked), and melee weapon injections can use the BlockSlots feature that was previously only on the OnCollide component. BlockSlots also didn't seem to be working; now it is.

Fixes #26233.
From what I can tell, also fixes #20708.
Provides one possible way of addressing #20699.

## Media

Code.
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

SolutionInjectOnCollideComponent and SolutionInjectOnCollideSystem have been removed.

Replace SolutionInjectOnCollide components in entity prototypes with SolutionInjectOnProjectileHit or SolutionInjectOnEmbed, depending on desired behavior.

**Changelog**

:cl:
- fix: spears, darts, and hypodarts can inject targets again.
- fix: arrows can no longer inject targets they don't actually hit.

### CODE REVIEW

**Comment by CatMagic1** [comment]
hallelua

File: Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        // Build a list of bloodstreams to inject into
+        var targetBloodstreams = new List<Entity<BloodstreamComponent>>();
+        foreach (var target in targets)
+        {
+            if (Deleted(target))
+                continue;
+
+            // Yuck, this is way to hardcodey for my tastes
+            // TODO blocking injection with a hardsuit should probably done with a cancellable event or something
+            if (!injector.Comp.PierceArmor && _inventory.TryGetSlotEntity(target, "outerClothing", out var suit) && _tag.HasTag(suit.Value, "Hardsuit"))
```
Comment (Tayrtahn) [inline Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs:88]: Like the comment says, this is kinda gross. I'm just carrying it over directly from the old melee injection system though; a future PR can replace it with something better.

File: Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                while (containerEnumerator.MoveNext(out var container))
+                {
+                    if (container.ContainedEntity != null)
+                    {
+                        blocked = true;
+                        break;
+                    }
+                }
+                if (blocked)
+                    continue;
```
Comment (Tayrtahn) [inline Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs:111]: The old logic in inject on collide was:
```c#
            if (containerEnumerator.MoveNext(out _)) 
                 return;
```
which didn't work at all? It just checked if the target had any of the slots, but didn't check if anything was actually in them.

File: Content.Server/Chemistry/Components/BaseSolutionInjectOnEventComponent.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    /// Passed values: $weapon and $target
+    /// </remarks>
+    [DataField]
+    public LocId BlockedByHardsuitPopupMessage = "melee-inject-failed-hardsuit";
+
+    /// <summary>
+    /// If anything covers any of these slots then the injection fails.
+    /// </summary>
+    [DataField]
+    public SlotFlags BlockSlots = SlotFlags.NONE;
```
Comment (Tayrtahn) [inline Content.Server/Chemistry/Components/BaseSolutionInjectOnEventComponent.cs:59]: This defaulted to MASK before, and then every entity prototype that used it set it to NONE except for one that set it to OUTERCLOTHING. So it defaults to NONE now.

File: Content.Server/Chemistry/Components/BaseSolutionInjectOnEventComponent.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float TransferEfficiency { get => _transferEfficiency; set => _transferEfficiency = Math.Clamp(value, 0, 1); }
+
+    /// <summary>
+    /// Proportion of the <see cref="TransferAmount"/> that will actually be injected
+    /// into the target's bloodstream. The rest is lost.
+    /// 0 means none of the transferred solution will enter the bloodstream.
+    /// 1 means the entire amount will enter the bloodstream.
+    /// </summary>
+    [DataField("transferEfficiency")]
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Chemistry/Components/BaseSolutionInjectOnEventComponent.cs:30]: It adds underscores to the field name itself.

[the remark this answers is not in the document]
Comment (Tayrtahn) [reply Content.Server/Chemistry/Components/BaseSolutionInjectOnEventComponent.cs:30]: Then explicitly labeled it shall remain

File: Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        // Make sure we have at least one target
+        if (targets.Count == 0)
+            return false;
+
+        // Get the solution to inject
+        if (!_solutionContainer.TryGetSolution(injector.Owner, injector.Comp.Solution, out var injectorSolution))
+            return false;
+
+        // Build a list of bloodstreams to inject into
+        var targetBloodstreams = new List<Entity<BloodstreamComponent>>();
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs:80]:
```suggestion

        // Build a list of bloodstreams to inject into
        var targetBloodstreams = new ValueList<Entity<BloodstreamComponent>>();
```

**Pull request merged**